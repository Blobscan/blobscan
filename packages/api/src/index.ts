import type { inferRouterInputs, inferRouterOutputs } from "@trpc/server";

import type { AppRouter } from "./app-router";

export { appRouter, type AppRouter } from "./app-router";
export { createTRPCContext, createTRPCInnerContext } from "./context";
export type {
  TRPCInnerContext as TRCInnerContext,
  TRPCContext,
} from "./context";

export * from "@trpc/server/adapters/express";

export { env } from "./env";
export { gracefulShutdown } from "./graceful-shutdown";
export { metricsHandler } from "./instrumentation";
export type {
  BlobStorage,
  Rollup,
  Metric,
  MetricHistogram,
  MetricHistogramBucket,
  Metrics,
  MetricsClient,
} from "@blobscan/db";

/**
 * Inference helpers for input types
 * @example type HelloInput = RouterInputs['example']['hello']
 **/
export type RouterInputs = inferRouterInputs<AppRouter>;

/**
 * Inference helpers for output types
 * @example type HelloOutput = RouterOutputs['example']['hello']
 **/
export type RouterOutputs = inferRouterOutputs<AppRouter>;
