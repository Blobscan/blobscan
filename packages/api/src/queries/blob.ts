import { Prisma } from "@blobscan/db";

export const blobSelect = Prisma.validator<Prisma.BlobSelect>()({
  versionedHash: true,
  commitment: true,
  size: true,
  dataStorageReferences: {
    select: {
      dataReference: true,
      blobStorage: true,
    },
  },
  transactions: {
    select: {
      index: true,
      txHash: true,
      transaction: {
        select: {
          blockNumber: true,
          block: {
            select: {
              timestamp: true,
            },
          },
        },
      },
    },
  },
});

export const blobsOnTransactionsSelect =
  Prisma.validator<Prisma.BlobsOnTransactionsSelect>()({
    blobHash: true,
    txHash: true,
    index: true,
    blob: {
      select: {
        commitment: true,
        size: true,
        dataStorageReferences: {
          select: {
            blobStorage: true,
            dataReference: true,
          },
        },
      },
    },
    transaction: {
      select: {
        blockNumber: true,
        block: {
          select: {
            timestamp: true,
          },
        },
      },
    },
  });
