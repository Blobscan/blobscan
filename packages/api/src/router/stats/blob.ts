import { z } from "zod";

import { timeFrameProcedure } from "../../middlewares/withTimeFrame";
import { createTRPCRouter, publicProcedure } from "../../trpc";
import { STATS_PATH } from "../../utils/stats";

export const blobStatsRouter = createTRPCRouter({
  getOverallStats: publicProcedure
    .meta({
      openapi: {
        method: "GET",
        path: `${STATS_PATH}/blob/overall`,
        tags: ["stats", "blob"],
        summary: "Get blob overall stats",
      },
    })
    .input(z.void())
    .output(
      z
        .object({
          totalBlobs: z.number(),
          totalUniqueBlobs: z.number(),
          totalBlobSize: z.bigint(),
          avgBlobSize: z.number(),
          updatedAt: z.date(),
        })
        .nullable()
    )
    .query(async ({ ctx: { prisma } }) =>
      prisma.blobOverallStats.findUnique({
        where: { id: 1 },
      })
    ),
  getDailyStats: timeFrameProcedure
    .meta({
      openapi: {
        method: "GET",
        path: `${STATS_PATH}/blob`,
        tags: ["stats", "blob"],
        summary: "Get blob time series stats",
      },
    })
    .output(
      z.array(
        z.object({
          day: z.date(),
          totalBlobs: z.number(),
          totalUniqueBlobs: z.number(),
          totalBlobSize: z.bigint(),
          avgBlobSize: z.number(),
        })
      )
    )
    .query(({ ctx: { prisma, timeFrame } }) =>
      prisma.blobDailyStats.findMany({
        select: {
          day: true,
          totalBlobs: true,
          totalUniqueBlobs: true,
          totalBlobSize: true,
          avgBlobSize: true,
        },
        where: {
          day: {
            gte: timeFrame.initial.toDate(),
            lte: timeFrame.final.toDate(),
          },
        },
        orderBy: { day: "asc" },
      })
    ),
});
