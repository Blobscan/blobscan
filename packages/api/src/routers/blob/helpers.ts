import type {
  BlobscanPrismaClient,
  EthUsdPrice,
  ExtendedBlobDataStorageReferenceSelect,
  Prisma,
} from "@blobscan/db";
import { EthUsdPriceModel } from "@blobscan/db/prisma/zod";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedBlock,
  ExpandedTransaction,
  Expands,
} from "../../middlewares/withExpands";
import type { Prettify } from "../../types";
import { isFullyDefined } from "../../utils";
import {
  normalizePrismaBlobDataStorageReferencesFields,
  normalizePrismaTransactionFields,
} from "../../utils";
import {
  baseBlobSchema,
  baseBlockSchema,
  baseTransactionSchema,
  prismaBlobOnTransactionSchema,
} from "../../zod-schemas";

const blobDataStorageReferenceSelect = {
  blobStorage: true,
  url: true,
} satisfies ExtendedBlobDataStorageReferenceSelect;

export const extendedBlobDataStorageReferenceArgs: Prisma.Args<
  BlobscanPrismaClient["blobDataStorageReference"],
  "findFirstOrThrow"
> = {
  select: blobDataStorageReferenceSelect,
  orderBy: {
    blobStorage: "asc",
  },
};

export const baseBlobSelect = {
  commitment: true,
  proof: true,
  size: true,
  usageSize: true,
  versionedHash: true,
} satisfies Prisma.BlobSelect;

export const baseBlobOnTransactionSelect = {
  blobHash: true,
  blockHash: true,
  blockNumber: true,
  blockTimestamp: true,
  index: true,
  txHash: true,
  txIndex: true,
} satisfies Prisma.BlobsOnTransactionsSelect;

export type ExtendedBlobDataStorageReference = NonNullable<
  Prisma.Result<
    BlobscanPrismaClient["blobDataStorageReference"],
    { select: typeof blobDataStorageReferenceSelect },
    "findFirst"
  >
>;
type PrismaBlob = Prisma.BlobGetPayload<{
  select: typeof baseBlobSelect;
}> & {
  dataStorageReferences: ExtendedBlobDataStorageReference[];
};

type PrismaBlobOnTransaction = Prisma.BlobsOnTransactionsGetPayload<{
  select: typeof baseBlobOnTransactionSelect;
}>;

export type CompletedPrismaBlob = Prettify<
  PrismaBlob & {
    transactions: Prettify<
      PrismaBlobOnTransaction & {
        block?: ExpandedBlock;
        transaction?: ExpandedTransaction & {
          block: { blobGasPrice: ExpandedBlock["blobGasPrice"] };
        };
      }
    >[];
  }
>;

export type CompletedPrismaBlobOnTransaction = Prettify<
  PrismaBlobOnTransaction & {
    blob: {
      dataStorageReferences: ExtendedBlobDataStorageReference[];
    } & ExpandedBlob;
    block?: ExpandedBlock;
    transaction?: ExpandedTransaction & {
      block: { blobGasPrice: ExpandedBlock["blobGasPrice"] };
    };
  }
>;
export function createBlobSelect(expands: Expands) {
  const blockExpand = expands.block ? { block: expands.block } : {};
  const txExpand = expands.transaction
    ? {
        transaction: {
          select: {
            ...expands.transaction.select,
            block: {
              select: {
                blobGasPrice: true,
              },
            },
          },
        },
      }
    : {};

  return {
    ...baseBlobSelect,
    dataStorageReferences: extendedBlobDataStorageReferenceArgs,
    transactions: {
      select: {
        ...baseBlobOnTransactionSelect,
        ...blockExpand,
        ...txExpand,
      },
      orderBy: {
        blockNumber: "desc",
      },
    },
  } satisfies Prisma.BlobSelect;
}

export function createBlobsOnTransactionsSelect(expands: Expands) {
  const blockExpand = expands.block ? { block: expands.block } : {};
  const txExpand = expands.transaction
    ? {
        transaction: {
          select: {
            ...expands.transaction.select,
            block: {
              select: {
                blobGasPrice: true,
              },
            },
          },
        },
      }
    : {};

  return {
    ...baseBlobOnTransactionSelect,
    blob: {
      select: {
        ...baseBlobSelect,
        dataStorageReferences: extendedBlobDataStorageReferenceArgs,
      },
    },
    ...txExpand,
    ...blockExpand,
  } satisfies Prisma.BlobsOnTransactionsSelect;
}

const responseBlobOnTxSchema = prismaBlobOnTransactionSchema
  .omit({ blobHash: true })
  .required({
    blockHash: true,
    blockNumber: true,
    blockTimestamp: true,
    index: true,
    txHash: true,
    txIndex: true,
  });

export const responseBlobSchema = baseBlobSchema.extend({
  transactions: z.array(
    responseBlobOnTxSchema
      .merge(
        baseTransactionSchema
          .omit({
            hash: true,
            blockNumber: true,
            blockTimestamp: true,
            blockHash: true,
          })
          .partial()
      )
      .extend({
        block: baseBlockSchema
          .omit({
            hash: true,
            number: true,
            timestamp: true,
          })
          .partial({
            blobGasBaseUsdFee: true,
            blobGasUsdPrice: true,
          })
          .optional(),
        ethUsdPrice: EthUsdPriceModel.shape.price.optional(),
      })
  ),
});

export const responseBlobOnTransactionSchema = responseBlobOnTxSchema
  .merge(baseBlobSchema)
  .extend({
    block: baseBlockSchema
      .omit({
        hash: true,
        number: true,
        timestamp: true,
      })
      .optional(),
    transaction: baseTransactionSchema
      .omit({
        hash: true,
        blockNumber: true,
        blockTimestamp: true,
        blockHash: true,
      })
      .optional(),
  });

export type ResponseBlob = z.input<typeof responseBlobSchema>;
export type ResponseBlobOnTransaction = z.input<
  typeof responseBlobOnTransactionSchema
>;

export function toResponseBlob(
  {
    dataStorageReferences,
    transactions: blobOnTxs,
    ...prismaBlob
  }: CompletedPrismaBlob,
  ethUsdPrices: EthUsdPrice["price"][]
): ResponseBlob {
  const responseBlob: ResponseBlob = {
    ...prismaBlob,
    dataStorageReferences: normalizePrismaBlobDataStorageReferencesFields(
      dataStorageReferences
    ),
    transactions: blobOnTxs.map(
      (
        {
          block: { computeUsdFields, ...restBlock } = {},
          transaction: prismaTx = {},
          ...restBlobOnTx
        },
        i
      ) => {
        const txEthUsdPrice = ethUsdPrices[i];

        const blockUsdFields =
          txEthUsdPrice && computeUsdFields
            ? computeUsdFields(txEthUsdPrice)
            : undefined;
        const txBlock = {
          ...restBlock,
          ...(blockUsdFields ?? {}),
        };

        let responseBlobOnTx: ResponseBlob["transactions"][number] = {
          ...restBlobOnTx,
          ...txBlock,
          ethUsdPrice: txEthUsdPrice?.toNumber(),
        };

        if (isFullyDefined(prismaTx)) {
          const {
            block,
            decodedFields,
            from,
            fromId,
            toId,
            computeBlobGasBaseFee,
            computeUsdFields,
            ...restPrismaTx
          } = prismaTx;

          responseBlobOnTx = {
            ...responseBlobOnTx,
            ...restPrismaTx,
            ...normalizePrismaTransactionFields({
              decodedFields,
              from,
              fromId,
              toId,
            }),
            blobGasBaseFee: computeBlobGasBaseFee(block.blobGasPrice),
            ...(txEthUsdPrice
              ? computeUsdFields({
                  blobGasPrice: block.blobGasPrice,
                  ethUsdPrice: txEthUsdPrice,
                })
              : {}),
          };
        }

        return responseBlobOnTx;
      }
    ),
  };

  return responseBlob;
}

export function toResponseBlobOnTransaction(
  prismaBlobOnTransaction: CompletedPrismaBlobOnTransaction,
  ethUsdPrice?: EthUsdPrice["price"]
): ResponseBlobOnTransaction {
  const {
    blobHash,
    blob: { dataStorageReferences, ...blob },
    transaction,
    block,
    ...restBlobOnTransaction
  } = prismaBlobOnTransaction;

  const responseBlobOnTx: ResponseBlobOnTransaction = {
    ...blob,
    versionedHash: blobHash,
    dataStorageReferences: normalizePrismaBlobDataStorageReferencesFields(
      dataStorageReferences
    ),
    ...restBlobOnTransaction,
  };

  if (transaction) {
    const {
      block: txBlock,
      decodedFields,
      from,
      fromId,
      toId,
      computeBlobGasBaseFee,
      computeUsdFields,
      ...restTx
    } = transaction;

    responseBlobOnTx.transaction = {
      ...restTx,
      blobGasPrice: txBlock.blobGasPrice,
      ...normalizePrismaTransactionFields({
        decodedFields,
        from,
        fromId,
        toId,
      }),
      blobGasBaseFee: computeBlobGasBaseFee(txBlock.blobGasPrice),
      ...(ethUsdPrice
        ? computeUsdFields({
            blobGasPrice: txBlock.blobGasPrice,
            ethUsdPrice,
          })
        : {}),
    };
  }

  if (block) {
    const { computeUsdFields, ...restBlock } = block;
    responseBlobOnTx.block = {
      ...restBlock,
      ...(ethUsdPrice ? computeUsdFields(ethUsdPrice) : {}),
    };
  }

  return responseBlobOnTx;
}
