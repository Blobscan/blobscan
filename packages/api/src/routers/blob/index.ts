import { t } from "../../trpc-client";
import { getAll } from "./getAll";
import { getByBlobId } from "./getByBlobId";

export const blobRouter = t.router({
  getAll,
  getByBlobId,
});
