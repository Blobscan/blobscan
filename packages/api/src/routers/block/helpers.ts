import type { BlobStorageManager } from "@blobscan/blob-storage-manager";
import type { BlobscanPrismaClient, Prisma } from "@blobscan/db";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedTransaction,
  Expands,
} from "../../middlewares/withExpands";
import type { Filters } from "../../middlewares/withFilters";
import type { Prettify } from "../../types";
import { isFullyDefined } from "../../utils";
import { retrieveBlobData } from "../../utils";
import {
  deriveTransactionFields,
  normalizePrismaBlobFields,
  normalizePrismaTransactionFields,
} from "../../utils/transformers";
import {
  baseBlobSchema,
  baseBlockSchema,
  baseTransactionSchema,
} from "../../zod-schemas";

const prismaBlockSelect = {
  hash: true,
  number: true,
  timestamp: true,
  slot: true,
  blobGasUsed: true,
  blobAsCalldataGasUsed: true,
  blobGasPrice: true,
  excessBlobGas: true,
} satisfies Prisma.BlockSelect;

export type PrismaBlock = Prisma.BlockGetPayload<{
  select: typeof prismaBlockSelect;
}>;

export type CompletePrismaBlock = Prettify<
  PrismaBlock & {
    transactions: Prettify<
      {
        hash: string;
      } & Partial<ExpandedTransaction> & {
          blobs: Prettify<
            { blobHash: string } & {
              blob?: ExpandedBlob;
            }
          >[];
        }
    >[];
  }
>;

export function createBlockSelect(expands: Expands, filters?: Filters) {
  const blobExpand = expands.blob ? { blob: expands.blob } : {};
  const transactionExpand = expands.transaction?.select ?? {};

  return {
    ...prismaBlockSelect,
    transactions: {
      select: {
        hash: true,
        ...transactionExpand,
        blobs: {
          select: {
            blobHash: true,
            ...blobExpand,
          },
          orderBy: {
            index: "asc",
          },
        },
      },
      orderBy: {
        index: "asc",
      },
      where: filters?.transactionFilters,
    },
  } satisfies Prisma.BlockSelect;
}

export const responseBlockSchema = baseBlockSchema.extend({
  transactions: z.array(
    baseTransactionSchema
      .omit({
        blockHash: true,
        blockNumber: true,
        blockTimestamp: true,
      })
      .partial()
      .required({ hash: true })
      .extend({
        blobs: z.array(
          baseBlobSchema.partial().required({ versionedHash: true })
        ),
      })
  ),
});

export type ResponseBlock = z.input<typeof responseBlockSchema>;

export function toResponseBlock(
  prismaBlock: CompletePrismaBlock
): ResponseBlock {
  const transactions = prismaBlock.transactions.map(
    ({ hash, blobs, ...prismaTx }) => ({
      hash,
      ...(isFullyDefined(prismaTx)
        ? {
            ...prismaTx,
            ...normalizePrismaTransactionFields(prismaTx),
            ...deriveTransactionFields({
              ...prismaTx,
              blobGasPrice: prismaBlock.blobGasPrice,
            }),
          }
        : {}),
      blobs: blobs.map(({ blobHash, blob }) =>
        blob
          ? normalizePrismaBlobFields({ versionedHash: blobHash, ...blob })
          : { versionedHash: blobHash }
      ),
    })
  );

  return {
    ...prismaBlock,
    transactions,
  };
}

export type BlockId = "hash" | "number" | "slot";
export type BlockIdField =
  | { type: "hash"; value: string }
  | { type: "number"; value: number }
  | { type: "slot"; value: number };

function buildBlockWhereClause(
  { type, value }: BlockIdField,
  filters: Filters
): Prisma.BlockWhereInput {
  switch (type) {
    case "hash": {
      return { hash: value };
    }
    case "number": {
      return { number: value, transactionForks: filters.blockType };
    }
    case "slot": {
      return { slot: value, transactionForks: filters.blockType };
    }
  }
}

export async function fetchBlock(
  blockId: BlockIdField,
  {
    blobStorageManager,
    prisma,
    filters,
    expands,
  }: {
    blobStorageManager: BlobStorageManager;
    prisma: BlobscanPrismaClient;
    filters: Filters;
    expands: Expands;
  }
): Promise<CompletePrismaBlock | undefined> {
  const select = createBlockSelect(expands);
  const where = buildBlockWhereClause(blockId, filters);

  const prismaBlock = (await prisma.block.findFirst({
    select,
    where,
  })) as unknown as CompletePrismaBlock | null;

  if (!prismaBlock) {
    return;
  }

  if (expands.blobData) {
    await Promise.all(
      prismaBlock.transactions
        .flatMap((tx) => tx.blobs)
        .map(async ({ blobHash, blob }) => {
          const dataStorageReferences = blob?.dataStorageReferences;
          if (dataStorageReferences) {
            const data = await retrieveBlobData(blobStorageManager, {
              dataStorageReferences,
              versionedHash: blobHash,
            });

            blob.data = data;
          }
        })
    );
  }

  return prismaBlock;
}
