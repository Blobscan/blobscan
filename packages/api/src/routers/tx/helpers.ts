import type { Prisma } from "@blobscan/db";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedBlock,
  Expands,
} from "../../middlewares/withExpands";
import type { Prettify } from "../../types";
import {
  deriveTransactionFields,
  normalizePrismaBlobFields,
  normalizePrismaTransactionFields,
} from "../../utils/transformers";
import {
  baseTransactionSchema,
  baseBlobSchema,
  baseBlockSchema,
} from "../../zod-schemas";

const transactionSelect = {
  hash: true,
  fromId: true,
  toId: true,
  blobGasUsed: true,
  blobAsCalldataGasUsed: true,
  gasPrice: true,
  maxFeePerBlobGas: true,
  blockHash: true,
  blockNumber: true,
  blockTimestamp: true,
  index: true,
  decodedFields: true,
  from: {
    select: {
      address: true,
      rollup: true,
    },
  },
} satisfies Prisma.TransactionSelect;

type PrismaTransaction = Prisma.TransactionGetPayload<{
  select: typeof transactionSelect;
}>;

export type CompletePrismaTransaction = Prettify<
  PrismaTransaction & {
    decodedFields: NonNullable<PrismaTransaction["decodedFields"]>;
    block: Prettify<
      Partial<ExpandedBlock> & {
        blobGasPrice: ExpandedBlock["blobGasPrice"];
      }
    >;
    blobs: Prettify<
      { blobHash: string } & {
        blob?: ExpandedBlob;
      }
    >[];
  }
>;
export function createTransactionSelect(expands: Expands) {
  const blockExpand = expands.block?.select;
  const blobExpand = expands.blob;

  return {
    ...transactionSelect,
    block: {
      select: {
        ...(blockExpand ?? {}),
        blobGasPrice: true,
      },
    },
    blobs: {
      select: {
        blobHash: true,
        ...(blobExpand ? { blob: blobExpand } : {}),
      },
      orderBy: {
        index: "asc",
      },
    },
  } satisfies Prisma.TransactionSelect;
}

export const responseTransactionSchema = baseTransactionSchema.extend({
  block: baseBlockSchema
    .omit({
      hash: true,
      number: true,
      timestamp: true,
    })
    .partial()
    .required({
      blobGasPrice: true,
    }),
  blobs: z.array(baseBlobSchema.partial().required({ versionedHash: true })),
});

export type ResponseTransaction = z.input<typeof responseTransactionSchema>;

export function toResponseTransaction(
  prismaTx: CompletePrismaTransaction
): ResponseTransaction {
  const { blobs: blobsOnTxs, block } = prismaTx;
  const normalizedFields = normalizePrismaTransactionFields(prismaTx);
  const derivedFields = deriveTransactionFields({
    ...prismaTx,
    blobGasPrice: block.blobGasPrice,
  });
  const blobs = blobsOnTxs.map(({ blobHash, blob }) =>
    blob
      ? normalizePrismaBlobFields({ versionedHash: blobHash, ...blob })
      : { versionedHash: blobHash }
  );

  return {
    ...prismaTx,
    ...normalizedFields,
    ...derivedFields,
    blobs,
  };
}
