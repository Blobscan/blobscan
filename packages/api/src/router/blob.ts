import { TRPCError } from "@trpc/server";
import { z } from "zod";

import { paginatedProcedure } from "../middlewares/withPagination";
import { blobSelect, blobsOnTransactionsSelect } from "../queries/blob";
import { createTRPCRouter, publicProcedure } from "../trpc";

export const blobRouter = createTRPCRouter({
  getAll: paginatedProcedure.query(async ({ ctx }) => {
    const [blobs, totalBlobs] = await Promise.all([
      ctx.prisma.blob.findMany({
        select: { ...blobSelect },
        ...ctx.pagination,
      }),
      ctx.prisma.blob.count(),
    ]);

    return {
      blobs,
      totalBlobs,
    };
  }),
  getByIndex: publicProcedure
    .input(
      z.object({
        txHash: z.string(),
        index: z.number(),
      })
    )
    .query(async ({ ctx: { prisma, blobStorageManager }, input }) => {
      const { txHash, index } = input;
      const blobOnTransaction = await prisma.blobsOnTransactions.findUnique({
        select: blobsOnTransactionsSelect,
        where: { txHash_index: { txHash, index } },
      });

      if (!blobOnTransaction) {
        throw new TRPCError({
          code: "NOT_FOUND",
          message: `No blob with tx hash ${txHash} and index ${index}`,
        });
      }

      const { blob, blobHash, transaction } = blobOnTransaction;
      const blobReferences: Parameters<typeof blobStorageManager.getBlob> = [
        { storage: "google", reference: blob.gsUri },
      ];

      if (blob.swarmHash) {
        blobReferences.push({ storage: "swarm", reference: blob.swarmHash });
      }

      const blobData = await blobStorageManager.getBlob(...blobReferences);

      if (!blobData) {
        throw new TRPCError({
          code: "NOT_FOUND",
          message: `No blob data with gs uri ${blob.gsUri} or swarm hash ${blob.swarmHash}`,
        });
      }

      return {
        versionedHash: blobHash,
        txHash,
        index,
        blockNumber: transaction.blockNumber,
        timestamp: transaction.block.timestamp,
        commitment: blob.commitment,
        data: blobData.data.toString(),
        size: blob.size,
        swarmHash: blob.swarmHash,
      };
    }),
});
