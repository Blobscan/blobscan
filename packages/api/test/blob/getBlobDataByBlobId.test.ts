import { TRPCError } from "@trpc/server";
import fs from "fs";
import { beforeEach, describe, expect, it, vi } from "vitest";

import type { BlobDataStorageReference } from "@blobscan/db";
import type { BlobStorage } from "@blobscan/db/prisma/enums";
import { env, fixtures, testValidError } from "@blobscan/test";

import type { TRPCContext } from "../../src";
import { createBlobRouter } from "../../src/routers/blob";
import { bytesToHex, hexToBytes } from "../../src/utils";
import { createTestContext } from "../helpers";
import { blobIdSchemaTestsSuite } from "../test-suites/schemas";

function buildGoogleBlobDataUrl(dataReference: string) {
  return `${env.GOOGLE_STORAGE_API_ENDPOINT}/storage/v1/b/${
    env.GOOGLE_STORAGE_BUCKET_NAME
  }/o/${encodeURIComponent(dataReference)}?alt=media`;
}

describe("getBlobDataByBlobId", () => {
  const blobRouter = createBlobRouter();
  let blobDataRouter: ReturnType<typeof createBlobRouter>;
  let authorizedCtx: TRPCContext;
  let authorizedBlobDataCaller: ReturnType<typeof blobRouter.createCaller>;

  const versionedHash =
    "0x01f433be851da7e34bf14bf4f21b4c7db4b38afee7ec74d3c576fdce9f8f6734";
  const unprefixedBlobData = fixtures.blobDatas
    .find((d) => d.id === versionedHash)
    ?.data.toString("hex");
  const expectedBlobData = `0x${unprefixedBlobData}`;

  beforeEach(async () => {
    vi.resetModules();
    vi.unmock("@blobscan/test");

    blobDataRouter = createBlobRouter({
      blobDataProcedure: { enabled: true, protected: true },
    });
    authorizedCtx = await createTestContext({
      apiClient: "blob-data",
    });
    authorizedBlobDataCaller = createBlobRouter({
      blobDataProcedure: { enabled: true, protected: true },
    }).createCaller(authorizedCtx);
  });

  describe("when authorized", () => {
    it("should fetch data by versioned hash", async () => {
      const result = await authorizedBlobDataCaller.getBlobDataByBlobId({
        id: versionedHash,
      });

      expect(result).toEqual(expectedBlobData);
    });

    it("should fetch data by kzg commitment", async () => {
      const commitment =
        "0x8c5b4383c1db58dc3f615ee8a1fdeb2a1ad19d1f26d72119c23b36b5df30ea4be9d55ccb9254f7a7993d23a78bd858ce";

      const result = await authorizedBlobDataCaller.getBlobDataByBlobId({
        id: commitment,
      });

      expect(result).toEqual(expectedBlobData);
    });

    describe("when fetching blob data stored in different formats", () => {
      const blobHash =
        "0x01000000000000000000000000000000000000000000000000000000000000b1";
      const blobBinFileName = `${blobHash.slice(2)}.bin`;
      const blobTxtFileName = `${blobHash.slice(2)}.txt`;
      const blobData = "0x0e2e5a3a2011ad49f5055eb3227d66d5";
      function createBlobDataStorageRef(
        blobStorage: BlobStorage,
        extension: "bin" | "txt"
      ): BlobDataStorageReference {
        return {
          blobHash: blobHash,
          blobStorage: blobStorage,
          dataReference: `1/01/00/00/01000000000000000000000000000000000000000000000000000000000000b1.${extension}`,
        };
      }

      beforeEach(async () => {
        fs.writeFileSync(
          blobBinFileName,
          hexToBytes("0x0e2e5a3a2011ad49f5055eb3227d66d5")
        );
        fs.writeFileSync(blobTxtFileName, blobData, {
          encoding: "utf-8",
        });

        await authorizedCtx.prisma.blobDataStorageReference.create;

        return async () => {
          await Promise.all([
            fs.promises.unlink(blobBinFileName),
            fs.promises.unlink(blobTxtFileName),
          ]);
        };
      });

      beforeEach(async () => {
        await authorizedCtx.prisma.blobDataStorageReference.deleteMany({
          where: {
            blobHash: blobHash,
          },
        });
      });

      it("should fetch data stored as a binary", async () => {
        const gcsBinRef = createBlobDataStorageRef("GOOGLE", "bin");
        const gcsUrl = buildGoogleBlobDataUrl(gcsBinRef.dataReference);
        const response = await fetch(gcsUrl);

        if (!response.ok) {
          throw new Error(await response.text());
        }

        const blobBytes = await response.arrayBuffer();
        const expectedBlobData = bytesToHex(blobBytes);

        await authorizedCtx.prisma.blobDataStorageReference.create({
          data: gcsBinRef,
        });

        const result = await authorizedBlobDataCaller.getBlobDataByBlobId({
          id: gcsBinRef.blobHash,
        });

        expect(result).toEqual(expectedBlobData);
      });

      it("should fetch data stored as text", async () => {
        const gcsTxtRef = createBlobDataStorageRef("GOOGLE", "txt");

        const gcsUrl = buildGoogleBlobDataUrl(gcsTxtRef.dataReference);
        const response = await fetch(gcsUrl);

        if (!response.ok) {
          throw new Error(await response.text());
        }

        const expectedBlobData = await response.text();

        await authorizedCtx.prisma.blobDataStorageReference.create({
          data: gcsTxtRef,
        });

        const result = await authorizedBlobDataCaller.getBlobDataByBlobId({
          id: gcsTxtRef.blobHash,
        });

        expect(result).toEqual(expectedBlobData);
      });
    });

    it("should fetch data stored in Postgres", async () => {
      const blobHash =
        "0x01000000000000000000000000000000000000000000000000000000000000b3";
      const expectedBlobData = await authorizedCtx.prisma.blobData
        .findUnique({
          where: {
            id: blobHash,
          },
        })
        .then((r) => (r?.data ? bytesToHex(r.data) : undefined));

      const result = await authorizedBlobDataCaller.getBlobDataByBlobId({
        id: blobHash,
      });

      expect(result).toEqual(expectedBlobData);
    });

    testValidError(
      "should fail when the blob data wasn't retrieved from any of the storages",
      async () => {
        await authorizedCtx.prisma.blobDataStorageReference.update({
          data: {
            dataReference: "123131231231231231231",
          },
          where: {
            blobHash_blobStorage: {
              blobHash: versionedHash,
              blobStorage: "POSTGRES",
            },
          },
        });
        await authorizedCtx.prisma.blobDataStorageReference.create({
          data: {
            blobHash: versionedHash,
            blobStorage: "GOOGLE",
            dataReference: "123131231231231231231",
          },
        });

        await authorizedBlobDataCaller.getBlobDataByBlobId({
          id: versionedHash,
        });
      },
      TRPCError,
      {
        checkCause: true,
      }
    );

    testValidError(
      "should fail when no blob data is found for the provided id",
      async () => {
        // blobDataAuthorizedContext = await createTestContext({
        //   apiClient: "blob-data",
        // });
        // authorizedBlobDataCaller = blobRouter.createCaller(
        //   blobDataAuthorizedContext
        // );

        await authorizedBlobDataCaller.getBlobDataByBlobId({
          id: "0x0130c6c0b2ed8e4951560d6c996ccab18486de35aee7a9064c957605c80d90d1",
        });
      },
      TRPCError
    );

    blobIdSchemaTestsSuite(async (invalidBlobId) => {
      await authorizedBlobDataCaller.getBlobDataByBlobId({
        id: invalidBlobId,
      });
    });
  });

  it("should fail when calling procedure without auth", async () => {
    vi.mock("@blobscan/test", async (original) => {
      const mod = (await original()) as { env: Record<string, unknown> };
      return {
        ...mod,
        env: {
          ...mod.env,
          BLOB_DATA_API_KEY: "key_secret",
          BLOB_DATA_API_ENABLED: true,
        },
      };
    });

    const ctx = await createTestContext();
    const unauthorizedBlobDataCaller = blobDataRouter.createCaller(ctx);

    await expect(
      unauthorizedBlobDataCaller.getBlobDataByBlobId({
        id: versionedHash,
      })
    ).rejects.toThrow(new TRPCError({ code: "UNAUTHORIZED" }));
  });
});
