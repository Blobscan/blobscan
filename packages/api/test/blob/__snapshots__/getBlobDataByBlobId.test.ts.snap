// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getBlobDataByBlobId > when authorized > should fail when no blob data is found for the provided id 1`] = `"No blob data storage URLs found for blob with id '0x0130c6c0b2ed8e4951560d6c996ccab18486de35aee7a9064c957605c80d90d1'."`;

exports[`getBlobDataByBlobId > when authorized > should fail when the blob data wasn't retrieved from any of the storages 1`] = `"Failed to fetch data for blob with id '0x01f433be851da7e34bf14bf4f21b4c7db4b38afee7ec74d3c576fdce9f8f6734' from any of the storages"`;

exports[`getBlobDataByBlobId > when authorized > should fail when the blob data wasn't retrieved from any of the storages 2`] = `undefined`;

exports[`getBlobDataByBlobId > when authorized > when providing an invalid blob id > should fail when providing a hex string with invalid length 1`] = `
"[
  {
    \\"code\\": \\"custom\\",
    \\"message\\": \\"Invalid input length: must be a versioned hash (32 bytes) or a kzg commitment (48 bytes)\\",
    \\"fatal\\": true,
    \\"path\\": [
      \\"id\\"
    ]
  }
]"
`;

exports[`getBlobDataByBlobId > when authorized > when providing an invalid blob id > should fail when providing a non-hex string 1`] = `
"[
  {
    \\"code\\": \\"invalid_string\\",
    \\"message\\": \\"Invalid input: must be a valid hex string\\",
    \\"validation\\": \\"regex\\",
    \\"fatal\\": true,
    \\"path\\": [
      \\"id\\"
    ]
  }
]"
`;

exports[`getBlobDataByBlobId > when authorized > when providing an invalid blob id > should fail when providing a versioned hash with invalid prefix 1`] = `
"[
  {
    \\"code\\": \\"invalid_string\\",
    \\"message\\": \\"Invalid versioned hash: must start with \\\\\\"0x01\\\\\\"\\",
    \\"validation\\": \\"regex\\",
    \\"fatal\\": true,
    \\"path\\": [
      \\"id\\"
    ]
  }
]"
`;
